lib/main.js

```diff
@@ -47,108 +42,215 @@ module.exports.runSass = function (options, describe, it) {
   });
 };
 
-// parse True output CSS and return an array of test modules.
-var parse = module.exports.parse = function (css) {
-  var modules = [];
-  var ast = parseCss(css);
-
-  _.each(_.where(ast.stylesheet.rules, {type: 'rule'}), function (rule) {
-    var decls = _.where(rule.declarations, {type: 'declaration'});
-    var declPairs = _.zip(
-      _.pluck(decls, 'property'),
-      _.pluck(decls, 'value')
-    );
-    var declInfo = _.object(declPairs);
-    var curAssertion, declString;
-    _.each(rule.selectors, function (sel) {
-      var selInfo = parseSelector(sel);
-      if (!selInfo) {
-        throw new Error("Can't understand selector: " + sel);
-      }
-      var curModule = _.findWhere(modules, {module: selInfo.module});
-      if (!curModule) {
-        curModule = {module: selInfo.module, tests: []};
-        modules.push(curModule);
-      }
-      var curTest = _.findWhere(curModule.tests, {test: selInfo.test});
-      if (!curTest) {
-        curTest = {test: selInfo.test, assertions: []};
-        curModule.tests.push(curTest);
-      }
-      if (selInfo.output) {
-        curAssertion = _.findWhere(
-          curTest.assertions, {description: selInfo.assert});
-        if (!curAssertion) {
-          curAssertion = {
-            description: selInfo.assert,
-            assert: 'equal',
-          };
-          curTest.assertions.push(curAssertion);
-        }
-        declString = _.map(declPairs, function (pair) {
-          return pair[0] + ': ' + pair[1] + ';';
-        }).join(' ');
-        if (selInfo.type === 'input') {
-          curAssertion.returned = declString;
-        } else { // if (selInfo.type == 'expect') {
-          curAssertion.expected = declString;
-        }
-        if (curAssertion.returned && curAssertion.expected) {
-          curAssertion.passed = curAssertion.returned === curAssertion.expected;
-        }
-      } else {
-        curAssertion = {
-          assert: selInfo.assert,
-          description: (declInfo['-description'] || '""').slice(1, -1),
-          passed: declInfo['-result'] === 'PASS',
-        };
-        _.each(_.pairs(declInfo), function (pair) {
-          if (_.contains(['-expected--', '-returned--'], pair[0].slice(0, 11))) {
-            var attr = pair[0].slice(1, 9);
-            var type = pair[0].slice(11);
-            var val = pair[1];
-            curAssertion[attr] = type + ': ' + val;
-          }
-        });
-        curTest.assertions.push(curAssertion);
-      }
-    });
+
+var parse = module.exports.parse = function (rawCss) {
+  var ast = css.parse(rawCss);
+  var ctx = { modules: [] };
+  var handler = parseModule;
+
+  _.each(ast.stylesheet.rules, function (rule) {
+    handler = handler(rule, ctx);
```

Comment:
Because the design here revolves around each handler returning the handler for whatever is expected next in the grammar. The outer definition is just the initial handler. Do you have an alternate implementation proposal?